File: tilde/aseipp/cs2wasm/tests/Features.cs
// SPDX-FileCopyrightText: © 2026 Austin Seipp
// SPDX-License-Identifier: Apache-2.0

namespace Tests;

public enum Kind : byte
{
    None,
    Paddle,
    Ball,
    Brick,
}

[System.Flags]
public enum Buttons : uint
{
    None = 0,
    Up = 1,
    Down = 2,
    Left = 4,
    Right = 8,
    A = 16,
}

public enum Level
{
    Debug,
    Info,
    Warn,
    Error,
}

internal static class Host
{
    [Gameplay.WasmImport("features", "read")]
    internal static extern int Read();
}

// Static initializers run on the first exported entry. While the host answers
// zero, the division faults and the next entry starts initialization again.
public static class Lazy
{
    static int hostValue = Host.Read();
    static int inverse = 100 / hostValue;

    public static int Inverse() => inverse;
}

// Static state persists across exported entries.
public static class Counter
{
    static int calls;
    static readonly int[] history = new int[4];
    static int seed = Seed();

    static int Seed() => 7;

    public static int Count { get; set; }

    public static int Calls => calls;

    public static int CallCount() => Calls * 100 + Count;

    public static int Next()
    {
        calls++;
        history[calls & 3] = calls * seed;
        Count += calls;
        return history[calls & 3];
    }
}

// Field initializers run in declaration order, then the static constructor.
public static class InitOrder
{
    static int a = 1;
    static int b = a + 10;
    static int c;

    static InitOrder()
    {
        c = b * 2;
    }

    public static int Sum() => a + b + c;
}

// A static constructor runs after every static field initializer of its
// class, even one declared after it.
public static class StaticConstructorOrder
{
    static StaticConstructorOrder()
    {
        x = x + 1;
        log = log * 10 + 3;
    }

    static int x = 5;
    static int log = 1;
    static int late = log * 10 + 2;

    public static int Get() => x;

    public static int Log() => log * 100 + late;
}

public static class Switches
{
    public static int Classify(int x)
    {
        switch (x)
        {
            case 0:
                return 100;
            case 1:
            case 2:
                return 200;
            default:
                return 900;
            case 3:
            {
                int y = x * 2;
                return y + 300;
            }
            case > 100 when x < 200:
                return 400;
            case < 0:
                return 500;
        }
    }

    public static int SwitchBreak(int n)
    {
        int total = 0;
        for (int i = 0; i < n; i++)
        {
            switch (i % 3)
            {
                case 0:
                    total += 1;
                    break;
                case 1:
                    continue;
                default:
                    total += 10;
                    break;
            }

            total += 100;
        }

        return total;
    }

    public static int SharedDefault(int x)
    {
        switch (x)
        {
            case 1:
                return 10;
            case 2:
            default:
                return 20;
            case 3:
                return 30;
        }
    }

    public static int KindScore(Kind kind) => kind switch
    {
        Kind.Paddle => 1,
        Kind.Ball or Kind.Brick => 2,
        _ => 0,
    };

    // No arm for values past Error: a fault, as the CLR's SwitchExpressionException.
#pragma warning disable CS8524
    public static int LevelWeight(int level) => (Level)level switch
    {
        Level.Debug => 1,
        Level.Info => 2,
        Level.Warn => 4,
        Level.Error => 8,
    };
#pragma warning restore CS8524

    public static int IsPattern(int x) => x is > 10 and < 20 ? 1 : (x is 5 or 6 ? 2 : 0);

    public static bool NotNull(bool useNull)
    {
        Body body = useNull ? null : new Body(1, 2);
        return body is not null;
    }

    public static int DoWhile(int n)
    {
        int i = 0;
        int sum = 0;
        do
        {
            i++;
            if (i == 2)
                continue;
            sum += i;
        }
        while (i < n);
        return sum;
    }

    // The switch block is one scope: a local declared in the default
    // section is assigned and used in a later section.
    public static int DefaultDeclares(int x)
    {
        switch (x)
        {
            default:
                int y = x + 1;
                return y;
            case 2:
                y = 5;
                return y * 10;
            case 3:
                Body body = new Body(x);
                return body.Y;
        }
    }

    public static double NaNCase(double x)
    {
        switch (x)
        {
            case double.NaN:
                return 1;
            case 0:
                return 2;
            default:
                return 3;
        }
    }
}

public static class Flags
{
    public static Buttons Press(Buttons state, Buttons button) => state | button;

    public static bool Held(Buttons state, Buttons button) => (state & button) == button && button != Buttons.None;

    public static Buttons Toggle(Buttons state, Buttons button) => state ^ button;

    public static Buttons Release(Buttons state, Buttons button) => state & ~button;

    // Kind is a byte enum: arithmetic wraps to the underlying type.
    public static Kind NextKind(Kind kind) => kind + 1;

    public static int Underlying(Kind kind) => (int)kind * 100 + (byte)kind;
}

public sealed class Body
{
    private int scale = 2;

    public Body(int x, int y)
    {
        X = x;
        Y = y;
    }

    public Body(int x) : this(x, x * 10)
    {
    }

    public int X { get; set; }

    public int Y { get; private set; }

    public int Sum => X + Y;

    public int Scaled
    {
        get => X * scale;
        set
        {
            X = value / scale;
        }
    }

    public int Fixed { get; } = 42;

    public int Clamped
    {
        get => field;
        set => field = value < 0 ? 0 : value;
    }

    public int Floor
    {
        get;
        set => field = value < 0 ? 0 : value;
    }

    public void Move(int dx)
    {
        X += dx;
    }
}

public static class Properties
{
    public static int Basic()
    {
        var body = new Body(3) { X = 5 };
        body.X++;
        body.Scaled += 4;
        body.Move(1);
        return body.Sum * 1000 + body.Fixed;
    }

    public static int Backed(int value)
    {
        var body = new Body(0);
        body.Clamped = value;
        body.Floor = value;
        body.Floor += 1;
        return body.Clamped * 100 + body.Floor;
    }

    public static int Discard()
    {
        var body = new Body(1);
        _ = body.Scaled;
        _ = new Body(2) { X = 3 };
        return body.X;
    }

    public static int Coalesce(bool useNull)
    {
        Body first = useNull ? null : new Body(1, 2);
        Body second = first ?? new Body(7, 8);
        return second.X;
    }

    public static int NullProperty()
    {
        Body body = null;
        return body.Sum;
    }
}

public static partial class Game
{
    [Gameplay.WasmExport("tick")]
    public static partial int Tick(int dt);

    public static class Inner
    {
        public sealed class Node
        {
            public int Value;
            public Node Next;
        }

        public static int Chain(int n)
        {
            Node head = null;
            for (int i = 1; i <= n; i++)
                head = new Node { Value = i, Next = head };
            int sum = 0;
            for (var node = head; node != null; node = node.Next)
                sum += node.Value;
            return sum;
        }
    }
}

public static partial class Game
{
    static int elapsed;

    public static partial int Tick(int dt)
    {
        elapsed += dt;
        return elapsed;
    }
}

internal static class Hidden
{
    public static int NotExported() => 1;
}

public static class Exposed
{
    internal static int AlsoNotExported() => 2;

    public static int Visible() => 3 + Hidden.NotExported() * 0 + AlsoNotExported() * 0;

    public static Body NotScalar() => null;
}

public static class Numbers
{
    public static long Wide(long a, long b) => a * b + (a >> 3) - (b >>> 1);

    public static ulong Unsigned(ulong a, uint b) => a / b + a % b;

    public static uint UnsignedDivide(uint a, uint b) => a / b;

    public static bool UnsignedLess(uint a, uint b) => a < b;

    public static int Saturate(double value) => (int)value;

    public static uint SaturateUnsigned(float value) => (uint)value;

    public static byte ToByte(int value) => (byte)value;

    public static sbyte ToSByte(int value) => (sbyte)value;

    public static short Shorten(long value) => (short)value;

    public static char NextChar(char character) => (char)(character + 1);

    public static int CharCode(char character) => character;

    public static byte ByteWrap(byte value)
    {
        value++;
        value += 250;
        return value;
    }

    public static long Extend(int value) => value;

    public static ulong ExtendUnsigned(uint value) => value;

    public static long ShiftWide(long value, int count) => value << count;

    public static double Mixed(float single, long wide) => single + wide;

    public static float FromUnsignedLong(ulong value) => value;

    public static int Absolute(int value) => System.Math.Abs(value);

    public static int Clamp(int value, int minimum, int maximum) => System.Math.Clamp(value, minimum, maximum);

    public static double Round(double value) => System.Math.Round(value);

    public static float Sqrt(float value) => System.MathF.Sqrt(value);

    public static double Max(double left, double right) => System.Math.Max(left, right);

    public static int PopCount(uint value) => System.Numerics.BitOperations.PopCount(value);

    public static uint Rotate(uint value, int count) => System.Numerics.BitOperations.RotateLeft(value, count);

    public static bool IsNaN(double value) => double.IsNaN(value);

    public static bool IsFinite(float value) => float.IsFinite(value);
}

File: tilde/aseipp/cs2wasm/tests/features.mjs
// SPDX-FileCopyrightText: © 2026 Austin Seipp
// SPDX-License-Identifier: Apache-2.0

// Statics, enums, switches, patterns, properties, partial classes, export
// naming and the wider numeric types. The module is instantiated here so the
// static-initialization retry can be driven through a host import.
import { checker } from "./checker.mjs";

export function checkFeatures(module, assert) {
  let hostValue = 0;
  const { exports } = new WebAssembly.Instance(module, {
    features: { read: () => hostValue },
  });
  const features = checker(exports, assert, "Tests.");
  const { equal, trap } = features;
  const call = (name, ...args) => exports["Tests." + name](...args);

  const exportNames = WebAssembly.Module.exports(module).map(({ name }) =>
    name
  );
  for (
    const name of [
      "tick",
      "Tests.Game.Inner.Chain",
      "Tests.Exposed.Visible",
      "Tests.Lazy.Inverse",
    ]
  ) {
    equal(exportNames.includes(name), true, `${name} is exported`);
  }
  for (
    const name of [
      "Tests.Game.Tick",
      "Tests.Hidden.NotExported",
      "Tests.Exposed.AlsoNotExported",
      "Tests.Exposed.NotScalar",
      "Tests.Host.Read",
    ]
  ) {
    equal(exportNames.includes(name), false, `${name} is not exported`);
  }

  // Static initialization runs on the first entry and is retried after a fault.
  trap("Lazy.Inverse", [], 7);
  trap("Numbers.Absolute", [1], 7);
  hostValue = 5;
  equal(
    call("Lazy.Inverse"),
    20,
    "initialization succeeds once the host cooperates",
  );
  equal(call("Counter.Next"), 7, "first call");
  equal(call("Counter.Next"), 14, "static state persists");
  equal(
    call("Counter.CallCount"),
    203,
    "static properties, bodied and automatic",
  );
  equal(
    call("InitOrder.Sum"),
    34,
    "initializer order, then the static constructor",
  );
  equal(
    call("StaticConstructorOrder.Get"),
    6,
    "static constructor after a later field initializer",
  );
  equal(
    call("StaticConstructorOrder.Log"),
    1312,
    "every field initializer, then the static constructor",
  );
  hostValue = 0;
  equal(call("Lazy.Inverse"), 20, "initialization does not run again");

  // Switch statements and expressions, patterns.
  for (
    const [input, expected] of [
      [0, 100],
      [1, 200],
      [2, 200],
      [3, 306],
      [150, 400],
      [250, 900],
      [-5, 500],
      [50, 900],
    ]
  ) {
    equal(call("Switches.Classify", input), expected, `Classify(${input})`);
  }
  equal(
    call("Switches.SwitchBreak", 6),
    422,
    "break and continue inside switch",
  );
  for (
    const [input, expected] of [
      [1, 10],
      [2, 20],
      [3, 30],
      [4, 20],
    ]
  ) {
    equal(
      call("Switches.SharedDefault", input),
      expected,
      `SharedDefault(${input})`,
    );
  }
  equal(call("Switches.KindScore", 1), 1, "enum switch expression");
  equal(call("Switches.KindScore", 3), 2, "or pattern");
  equal(call("Switches.KindScore", 257), 1, "byte enum argument canonicalized");
  equal(call("Switches.LevelWeight", 2), 4, "exhaustive by hand");
  trap("Switches.LevelWeight", [4], 11);
  equal(call("Switches.IsPattern", 15), 1, "relational and pattern");
  equal(call("Switches.IsPattern", 6), 2, "constant or pattern");
  equal(call("Switches.IsPattern", 20), 0, "no pattern");
  equal(call("Switches.NotNull", 0), 1, "is not null");
  equal(call("Switches.NotNull", 1), 0, "is not null on null");
  equal(call("Switches.DoWhile", 4), 8, "do/while with continue");
  equal(call("Switches.DoWhile", 0), 1, "do/while body runs once");
  for (
    const [input, expected] of [
      [7, 8],
      [2, 50],
      [3, 30],
    ]
  ) {
    equal(
      call("Switches.DefaultDeclares", input),
      expected,
      `DefaultDeclares(${input})`,
    );
  }
  equal(call("Switches.NaNCase", NaN), 1, "NaN constant pattern");
  equal(call("Switches.NaNCase", -0), 2, "negative zero matches zero");
  equal(call("Switches.NaNCase", 1), 3, "default");

  // Flags and narrow enums.
  equal(call("Flags.Press", 1, 8), 9, "or");
  equal(call("Flags.Held", 9, 8), 1, "and");
  equal(call("Flags.Held", 9, 0), 0, "None is never held");
  equal(call("Flags.Toggle", 9, 1), 8, "xor");
  equal(call("Flags.Release", 31, 2), 29, "and not");
  equal(call("Flags.NextKind", 3), 4, "enum arithmetic");
  equal(call("Flags.NextKind", 255), 0, "byte enum arithmetic wraps");
  equal(call("Flags.Underlying", 2), 202, "enum conversions");

  // Properties, constructor chaining, ?? and discards.
  equal(
    call("Properties.Basic"),
    39042,
    "auto, bodied and get-only properties",
  );
  equal(
    call("Properties.Backed", -5),
    1,
    "field keyword with clamping setters",
  );
  equal(call("Properties.Backed", 7), 708, "field keyword");
  equal(call("Properties.Discard"), 1, "discarded expressions still evaluate");
  equal(call("Properties.Coalesce", 0), 1, "?? keeps a non-null value");
  equal(call("Properties.Coalesce", 1), 7, "?? evaluates the fallback");
  trap("Properties.NullProperty", [], 5);

  // Partial classes and methods, nested classes, export naming.
  equal(exports.tick(16), 16, "WasmExport name");
  equal(exports.tick(16), 32, "partial method body sees static state");
  equal(
    call("Game.Inner.Chain", 4),
    10,
    "nested class with nested sealed class",
  );
  equal(call("Exposed.Visible"), 3, "public method of a public class");

  // Wider numeric types cross the boundary as i32/i64.
  equal(call("Numbers.Wide", 3n, 4n), 10n, "i64 arithmetic");
  equal(call("Numbers.Wide", -8n, 4n), -35n, "signed i64 shift");
  equal(
    call("Numbers.Unsigned", 2n ** 64n - 1n, 10),
    1844674407370955166n,
    "unsigned 64-bit division",
  );
  equal(
    call("Numbers.UnsignedDivide", -1, 2),
    2147483647,
    "unsigned 32-bit division",
  );
  equal(call("Numbers.UnsignedLess", -1, 1), 0, "unsigned comparison");
  equal(call("Numbers.Saturate", 1e10), 2147483647, "double to int saturates");
  equal(
    call("Numbers.Saturate", -1e10),
    -2147483648,
    "double to int saturates below",
  );
  equal(call("Numbers.Saturate", NaN), 0, "NaN to int is zero");
  equal(call("Numbers.Saturate", -2.9), -2, "truncation toward zero");
  equal(
    call("Numbers.SaturateUnsigned", -5),
    0,
    "float to uint saturates at zero",
  );
  equal(
    call("Numbers.SaturateUnsigned", 4e9),
    -294967296,
    "float to uint above int range",
  );
  equal(
    call("Numbers.SaturateUnsigned", 5e9),
    -1,
    "float to uint saturates above",
  );
  equal(call("Numbers.ToByte", 300), 44, "int to byte wraps");
  equal(call("Numbers.ToByte", -1), 255, "negative int to byte wraps");
  equal(call("Numbers.ToSByte", 200), -56, "int to sbyte wraps");
  equal(call("Numbers.Shorten", 70000n), 4464, "long to short wraps");
  equal(call("Numbers.NextChar", 65), 66, "char arithmetic");
  equal(call("Numbers.NextChar", 65535), 0, "char wraps");
  equal(call("Numbers.CharCode", 65601), 65, "char argument canonicalized");
  equal(
    call("Numbers.ByteWrap", 10),
    5,
    "byte increment and compound assignment wrap",
  );
  equal(call("Numbers.Extend", -1), -1n, "int to long sign-extends");
  equal(
    call("Numbers.ExtendUnsigned", -1),
    4294967295n,
    "uint to ulong zero-extends",
  );
  equal(
    call("Numbers.ShiftWide", 1n, 40),
    1099511627776n,
    "long shifted by an int",
  );
  equal(call("Numbers.ShiftWide", 1n, 65), 2n, "shift count masked");
  equal(call("Numbers.Mixed", 0.5, 2n), 2.5, "float and long widen to double");
  equal(
    call("Numbers.FromUnsignedLong", -1n),
    18446744073709552000,
    "ulong to float is unsigned",
  );
  equal(call("Numbers.Absolute", -7), 7, "Math.Abs");
  trap("Numbers.Absolute", [-2147483648], 9);
  equal(call("Numbers.Clamp", 15, 0, 10), 10, "Math.Clamp");
  equal(call("Numbers.Clamp", -3, 0, 10), 0, "Math.Clamp below");
  trap("Numbers.Clamp", [1, 5, 0], 10);
  equal(call("Numbers.Round", 2.5), 2, "Math.Round rounds half to even");
  equal(call("Numbers.Round", -3.5), -4, "Math.Round negative half");
  equal(call("Numbers.Sqrt", 2.25), 1.5, "MathF.Sqrt");
  equal(call("Numbers.Max", -0, 0), 0, "Math.Max orders signed zeros");
  equal(
    Object.is(call("Numbers.Max", -0, 0), 0),
    true,
    "Math.Max returns positive zero",
  );
  equal(
    Number.isNaN(call("Numbers.Max", NaN, 1)),
    true,
    "Math.Max propagates NaN",
  );
  equal(call("Numbers.PopCount", 0xf0f0), 8, "BitOperations.PopCount");
  equal(call("Numbers.Rotate", 0x80000001, 1), 3, "BitOperations.RotateLeft");
  equal(call("Numbers.IsNaN", NaN), 1, "double.IsNaN");
  equal(call("Numbers.IsNaN", 1), 0, "double.IsNaN on a number");
  equal(call("Numbers.IsFinite", Infinity), 0, "float.IsFinite on infinity");
  equal(call("Numbers.IsFinite", 1.5), 1, "float.IsFinite");
  return features.checks;
}
